### ISSUE TITLE
Optional chaining parentheses semantics lost

### ISSUE THREAD

**Issue by evanw** [microsoft/TypeScript#50148]
# Bug Report

I came across https://github.com/swc-project/swc/issues/5332 and figured that I should report this here, since I don't see that represented in this issue tracker.

### 🔎 Search Terms

["optional chaining" parentheses](https://github.com/microsoft/TypeScript/issues?q=is%3Aissue+is%3Aopen+%22optional+chaining%22+parentheses)

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, including nightly

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=9&ts=4.8.0-dev.20220802#code/CYUwxgNghgTiAEEQBd5QFzwK4DtQDMBLHEYeAH3gG94AjTAZ2RmIHMBtAXXgF8AoJKnwB7YfAC88ABRQA-ADpaaBvCYscHTgEp5SDcgAWAlHVgS0C2rpD6jg+GAPgA1gH0RwzMgCeABxDC+PAe8CAAHsg2wCo4WAC2tCAw8LLwzFgImPhQEAwIkukgxqiOLq60sF5+AUEVyeGReDHxicmphfBZOXnm2blFfAD0AFR8yLCsKJgAogDKAEwADACMAOzFwaLmMhLikrEQEBSUULuSAG7ChGSLKfCX1-C3mFCKOnqshgDcG3Xmp3t9lhDsc0Gd7lcbncHlCXoprLYfvZSmA3CECjAMkiTCi3H9JH08j9hoM+ENRuMYJNkDMFot5vMNuiLPCPt94INBvAADx7eAABQAggAlaYAOQAKgAJOZzeAi6bwAAyAHlZhL4DLRQBCX5mSRyVk2T4GbElJyo9xbDFYja48r64LdEDE0lAA)

### 💻 Code

```ts
declare let a: undefined | { b: string[] }
let foo = (a?.b as string[]).length
let bar = a?.b.length
let check_foo: typeof foo extends number ? true : false = true
let check_bar: typeof bar extends number ? true : false = false
```

### 🙁 Actual behavior

TypeScript generates this code with `target: "ES2017"`:

```js
let foo = (a === null || a === void 0 ? void 0 : a.b).length;
let bar = a === null || a === void 0 ? void 0 : a.b.length;
let check_foo = true;
let check_bar = false;
```

TypeScript generates this code with `target: "ES2022"`:

```js
let foo = a?.b.length; // <== PARENTHESES ARE LOST HERE!
let bar = a?.b.length;
let check_foo = true;
let check_bar = false;
```

### 🙂 Expected behavior

I expect TypeScript's generated output for `target: "ES2022"` to be this instead:

```js
let foo = (a?.b).length;
let bar = a?.b.length;
let check_foo = true;
let check_bar = false;
```

My rationale:

* That's how parentheses interact with optional chaining in JavaScript
* TypeScript's type system appears to think that `foo` cannot be `undefined` here, as demonstrated by the compile-time checks in the code
* TypeScript's code generation for `target: "ES2017"` seems to indicate that parentheses should be present with `target: "ES2022"` as well

I would like to know if this is considered a bug in TypeScript or if this is intended behavior. If this is intended behavior, then I'll have to update esbuild's TypeScript-to-JavaScript conversion code to match this behavior.

**Comment by MartinJohns** [comment]
> * TypeScript's type system appears to think that `foo` cannot be `undefined` here, as demonstrated by the compile-time checks in the code

**You** are telling TypeScripts type-system that it can't be `undefined`. Using the type assertion you say `a?.b` should be treated as `string[]`, and accessing `length` on `string[]` will never be `undefined`.

**Comment by RyanCavanaugh** [comment]
>  as (a?.b).length is the same as a?.b.length

This isn't true; when `a` is `undefined` the former will throw but the latter will not.

In other cases we've had logic that says an expression of the form `(a as b).c` can always emit as `a.b.c`; this is generally desirable for minification reasons but isn't safe to do here. We certainly shouldn't be varying the behavior based on target; that's a footgun.

**Comment by MartinJohns** [comment]
> This isn't true; when `a` is `undefined` the former will throw but the latter will not.

Could have sworn I tested this before commenting. I guess I shouldn't comment on issues so late at night.

**Comment by abrenneke** [comment]
Thanks for opening this @evanw I was considering opening a TS issue for that ticket!

> We certainly shouldn't be varying the behavior based on target; that's a footgun.

That's actually roughly what bit me in https://github.com/swc-project/swc/issues/5332 and led me to open the swc issue, actually!

For what it's worth someone on my team said that he likes `(a?.b as c).d` to be `a?.b.d` because it's "TypeScript knowing what I mean" (i.e. the cast requires parens syntactically but only casting shouldn't change semantics). The ES2017 code could be changed instead? But maybe that would be more confusing 🤷

**Comment by rbuckton** [comment]
Cast does not itself require parens, `const x = y as string` is perfectly legal. We are essentially erasing only the `as T` portion, so `(a as T)` should really just be `(a)`, but we have historically cleaned up the excess `()` to improve readability in the emit. However, `(a?.b).c` is semantically different from `a?.b.c` in JS so erasing the cast should *not* remove the parentheses. This is what #50156 addresses.
